lib/SILOptimizer/Analysis/MemoryBehavior.cpp

```diff
@@ -511,6 +514,28 @@ visitBeginCOWMutationInst(BeginCOWMutationInst *BCMI) {
   return MemBehavior::None;
 }
 
+MemBehavior MemoryBehaviorVisitor::visitLoadBorrowInst(LoadBorrowInst *lbi) {
+  if (!mayAlias(lbi->getOperand()))
+    return MemBehavior::None;
+  return MemBehavior::MayRead;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
+  return MemBehavior::None;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitEndBorrowInst(EndBorrowInst *ebi) {
+  auto borrowIntroducer = ebi->getOperand();
+  if (auto *lbi = dyn_cast<LoadBorrowInst>(borrowIntroducer)) {
+    if (!mayAlias(lbi->getOperand()))
+      return MemBehavior::None;
+  }
+  if (auto *bbi = dyn_cast<BeginBorrowInst>(borrowIntroducer)) {
+    return MemBehavior::None;
+  }
+  return MemBehavior::MayRead;
```

Comment:
Why do we have to treat an end_borrow as memory reading?

Reply:
end_borrow associated with a load_borrow is `MayRead`. When we can't cheaply find if the end_borrow is associated with a begin_borrow or load_borrow, I am returning `MayRead`.  
I did not want to write walking the def-use chain to find if the end_borrow was associated with a begin_borrow or load_borrow.

Reply:
@atrick can you look over `MemoryBehaviorVisitor::visitEndBorrowInst` and confirm the MemoryBehavior of end_borrow of load_borrow again ?

Reply:
I am having second thoughts about this. It seems like if I am marking an end_borrow of a load_borrow as MayRead to avoid any side-effects being moved into the borrow scope. I should also mark end_borrow of a begin_borrow as a MayRead for the same purpose.

Reply:
@eeckstein @meg-gupta The end borrow guarantees that the borrowed value will not be destroyed before that point. If the borrowed value was loaded from memory, then it's actually borrowing the in-memory reference. We can't let anyone write to that memory, which would release the in-memory reference early. At least that was the thinking. We should have a comment explaining that.

Reply:
Oh, I see. We define it as memory-read of the loaded value to prevent any other memory writes (to the same address) to be moved across the end_borrow.

That really needs to be explained in a comment. Please also mention in the comment that the last `return MemBehavior::MayRead` is a conservative default in case the begin-borrow can not be identified.

Reply:
Yes. @eeckstein my thinking is that with a sufficient analysis we should be able to always identify that. This at least is a good default until that is written.